src/librustc_metadata/creader.rs

```diff
@@ -854,21 +854,45 @@ impl<'a> CrateLoader<'a> {
                 return
             }
 
-            if !self.sess.crate_types.borrow().iter().all(|ct| {
-                match *ct {
-                    // Link the runtime
-                    config::CrateTypeExecutable => true,
-                    // This crate will be compiled with the required
-                    // instrumentation pass
-                    config::CrateTypeRlib => false,
-                    _ => {
-                        self.sess.err(&format!("Only executables and rlibs can be \
-                                                compiled with `-Z sanitizer`"));
-                        false
+            // The gnu linux target has better support for asan with multiple lib types.
```

Comment:
I don't have access to OSX so I can not debug the issues at hand here - my requirement is to add support for ASAN to dylibs so that I can start to use rust in another open source project. If I can't debug and isolate the OSX issue then the best answer for me is to avoid the problem and leave it to someone who has the tools and ability to isolate the root cause :(